File: ios/CCC/Analytics/AnalyticsManagerImpl.swift
//
//  AnalyticsManagerImpl.swift
//  CCC
//
//  Created by Mustafa Ozhan on 10.08.22.
//  Copyright © 2022 orgName. All rights reserved.
//

import Provider
import FirebaseAnalytics

class AnalyticsManagerImpl: AnalyticsManager {

    func setUserProperty(userProperty: UserProperty) {
        Analytics.setUserProperty(userProperty.value, forName: userProperty.key)
    }

    func trackEvent(event: Event) {
        if event.getParams() == nil {
            Analytics.logEvent(event.key, parameters: nil)
        } else {

            var analyticsParams = [String: String]()
            for (key, value) in event.getParams()! {
                analyticsParams[key] = value
            }

            Analytics.logEvent(event.key, parameters: analyticsParams)
        }
    }

    func trackScreen(screenName: ScreenName) {
        Analytics.logEvent(
            AnalyticsEventScreenView,
            parameters: [
                AnalyticsParameterScreenName: screenName.getScreenName(),
                AnalyticsParameterScreenClass: screenName.getScreenName()
            ]
        )
    }
}

File: client/core/analytics/src/androidMain/kotlin/com/oztechan/ccc/client/core/analytics/AnalyticsManagerImpl.kt
package com.oztechan.ccc.client.core.analytics

import com.google.firebase.Firebase
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.analytics
import com.google.firebase.analytics.logEvent
import com.oztechan.ccc.client.core.analytics.model.Event
import com.oztechan.ccc.client.core.analytics.model.ScreenName
import com.oztechan.ccc.client.core.analytics.model.UserProperty

internal class AnalyticsManagerImpl : AnalyticsManager {
    private val firebaseAnalytics by lazy { Firebase.analytics }

    override fun trackScreen(screenName: ScreenName) {
        firebaseAnalytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW) {
            param(FirebaseAnalytics.Param.SCREEN_NAME, screenName.getScreenName())
            param(FirebaseAnalytics.Param.SCREEN_CLASS, screenName.getScreenName())
        }
    }

    override fun setUserProperty(userProperty: UserProperty) {
        firebaseAnalytics.setUserProperty(userProperty.key, userProperty.value)
    }

    override fun trackEvent(event: Event) {
        firebaseAnalytics.logEvent(event.key) {
            event.getParams()
                ?.iterator()
                ?.forEach {
                    param(it.key, it.value)
                }
        }
    }
}
